.github/actions/bump-version/action.yml

```diff
@@ -0,0 +1,149 @@
+      'calver-minor': 'YYYY.OM.MINOR'.
+      'semver-major-minor': 'MAJOR.MINOR'.
+  bump-type:
+    required: true
+    type: string
+    description: |
+      The type of bump. This changes depending on the `versioning-scheme`.
+      'calver-minor': 'major' or 'minor'.
+      'semver-major-minor': 'major' or 'minor'.
+outputs:
+  before:
+    description: Version before being bumped. Equivalent to the input `version`.
+    value: ${{ steps.results.outputs.before }}
+  after:
+    description: Version after being bumped.
+    value: ${{ steps.results.outputs.after }}
+  type:
+    description: Type of bump. Equivalent to the input `bump-type`.
+    value: ${{ steps.results.outputs.type }}
+runs:
+  using: composite
+  steps:
+    - name: Check Versioning Scheme Validity
+      shell: bash
+      env:
+        VALID_VERSIONING_SCHEMES: calver-minor semver-major-minor
+
+      run: |
+        # check validity of versioning schemes
+        found=false
+        for valid_scheme in ${{ env.VALID_VERSIONING_SCHEMES }}; do
+          if [[ "${{ inputs.versioning-scheme }}" == "${valid_scheme}" ]]; then
+            found=true
+            break
+          fi
+        done
+        if [[ "${found}" == "false" ]]; then
+          echo "::error::${{ inputs.versioning-scheme }} is not supported by this action. Please check the list of allowed 'versioning-scheme's."
+          exit 1
+        fi
+
+    - name: Check Bump Type Validity for ${{ inputs.versioning.scheme }}
+      shell: bash
+      env:
+        VALID_CALVER_MINOR_BUMP_TYPES: major minor
+        VALID_SEMVER_MAJOR_MINOR_BUMP_TYPES: major minor
+      run: |
+        # check validity of bump types for each versioning scheme
+        found=false
+
+        if [[ "${{ inputs.versioning-scheme }}" == "calver-minor" ]]; then
+          for valid_bump in ${{ env.VALID_CALVER_MINOR_BUMP_TYPES }}; do
+            if [[ "${{ inputs.bump-type }}" == "${valid_bump}" ]]; then
+              exit 0
+            fi
+          done
+        elif [[ "${{ inputs.versioning-scheme }}" == "semver-major-minor" ]]; then
+          for valid_bump in ${{ env.VALID_SEMVER_MAJOR_MINOR_BUMP_TYPES }}; do
+            if [[ "${{ inputs.bump-type }}" == "${valid_bump}" ]]; then
+              exit 0
+            fi
+          done
+        fi
+
+        echo "::error::Bump type ${{ inputs.bump-type }} is not supported in ${{ inputs.versioning-scheme }}. Please check the list of allowed 'bump-type's."
+        exit 1
+
+    - name: Bump - calver-minor
+      shell: bash
+      if: inputs.versioning-scheme == 'calver-minor'
+      id: calver-minor
+      # Regarding the regex in the script: `([0-9]{4}\.(0[1-9]|1[1-2]))\.([0-9]+)` is broken down into:
+      # `([0-9]{4}\.(0[1-9]|1[1-2]))`: Major version as a date in the form YYYY.0M (eg. `2024.03`)
+      # `\.`: Version separator (eg. `.`)
+      # `([0-9]+)`: Minor version (eg. `1`)
+      # which would give `2024.03.1`
```

Comment:
Putting the script below into a file as bash functions would allow for some testing of the testing infra. No idea how much of a pain this would be though.

Reply:
I really do see the value in this...The issue is where do we put scripts in a repository that contains many (unrelated) workflows without it becoming wildly confusing?

Reply:
Have a `script` or `tools` directory and sub-dirs under that named for the CI that calls the function? That isn't great for code re-use though, which should be a goal, though I don't know how realistic or likely code reuse might be.

As I hinted to above, but didn't say explicitly, `bash` supports functions, so it isn't necessary to have a file for every function and call a shell script. We could have a smaller number of files we [`source`](https://stackoverflow.com/questions/20094271/using-dot-or-source-while-calling-another-script-what-is-the-difference) and call those functions directly from within a `bash` step. So it's code-reuse, it's independently testable, but doesn't have all the rigmarole of argument passing (and parsing) and potentially is a bit neater.